packages/plugin-ext/src/plugin/preference-registry.ts

```diff
@@ -0,0 +1,183 @@
+ */
+
+import { Emitter, Event } from '@theia/core/lib/common/event';
+import * as theia from '@theia/plugin';
+import {
+    PLUGIN_RPC_CONTEXT,
+    PreferenceRegistryExt,
+    PreferenceRegistryMain
+} from '../api/plugin-api';
+import { RPCProtocol } from '../api/rpc-protocol';
+import { isObject } from '../common/types';
+import { deepClone } from '../common/objects';
+import { PreferenceChange } from '@theia/core/lib/browser';
+import { ConfigurationTarget } from './types-impl';
+
+interface ConfigurationInspect<T> {
+    key: string;
+    defaultValue?: T;
+    globalValue?: T;
+    workspaceValue?: T;
+    workspaceFolderValue?: T;
+}
+
+function lookUp(tree: any, key: string): any {
+    if (key) {
+        const parts = key.split('.');
+        let node = tree;
+        for (let i = 0; node && i < parts.length; i++) {
+            node = node[parts[i]];
+        }
+        return node;
+    }
+}
+
+export class PreferenceRegistryExtImpl implements PreferenceRegistryExt {
+    private proxy: PreferenceRegistryMain;
+    private _preferences: any;
+    private readonly _onDidChangeConfiguration = new Emitter<theia.ConfigurationChangeEvent>();
+
+    readonly onDidChangeConfiguration: Event<theia.ConfigurationChangeEvent> = this._onDidChangeConfiguration.event;
+
+    constructor(rpc: RPCProtocol) {
+        this.proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.PREFERENCE_REGISTRY_MAIN);
+    }
+
+    $acceptConfigurationChanged(data: { [key: string]: any }, eventData: PreferenceChange): void {
+        this._preferences = this._parse(data);
+        this._onDidChangeConfiguration.fire(this._toConfigurationChangeEvent(eventData));
+    }
+
+    getConfiguration(section?: string, resource?: theia.Uri | null, extensionId?: string): theia.WorkspaceConfiguration {
+        const preferences = this._toReadonlyValue(section
+            ? lookUp(this._preferences, section)
+            : this._preferences);
+
+        const configuration: theia.WorkspaceConfiguration = {
+            has(key: string): boolean {
+                return typeof lookUp(preferences, key) !== 'undefined';
+            },
+            get: <T>(key: string, defaultValue?: T) => {
+                const result = lookUp(preferences, key);
+                if (typeof result === 'undefined') {
+                    return defaultValue;
+                } else {
+                    let clonedConfig: any = undefined;
+                    const cloneOnWriteProxy = (target: any, accessor: string): any => {
+                        let clonedTarget: any = undefined;
+                        const cloneTarget = () => {
+                            clonedConfig = clonedConfig ? clonedConfig : deepClone(preferences);
+                            clonedTarget = clonedTarget ? cloneTarget : lookUp(clonedConfig, accessor);
+                        };
+                        return isObject(target)
+                            ? new Proxy(target, {
```

Comment:
I would prefer not to use ternary operator here, its hard to read

Reply:
OK, I'll refactor this piece of code